### PR TITLE
Make sync_trait weakref callback into function (#69)

### PR DESCRIPTION
**Pull request by jvkersch**
This is a quick fix for #69. Essentially, the fix is to make the weakref callback for the synced objects into a local function instead of a `HasTraits` member.

This PR includes #134, since fixing the reference cycles meant that the issue in #131 also started appearing for mutual syncing.

### CODE REVIEW

**Comment by dpinte** [comment]
@jvkersch LGTM!

File: traits/has_traits.py [hunk 1]
```diff
@@ -2649,10 +2649,31 @@ def sync_trait ( self, trait_name, object, alias = None, mutual = True,
...
             return
 
-        value = ( weakref.ref( object, self._sync_trait_listener_deleted ),
-                  alias )
-        dic   = self._get_sync_trait_info().setdefault( trait_name, {} )
+        # Callback to use when the synced object goes out of scope. In order
+        # to avoid reference cycles, this must not be a member function. See
+        # Github issue #69 for more detail.
+
+        # FIXME: This callback removes the synced object from the
```
Comment (mdickinson) [inline traits/has_traits.py]: Is there an issue open for this?  If so, please reference the issue from the comment.  If not, please make one. :-)

File: traits/has_traits.py [hunk 2]
```diff
@@ -2649,10 +2649,31 @@ def sync_trait ( self, trait_name, object, alias = None, mutual = True,
...
+        # `_sync_trait_modified` trait listener if there are no more objects to
+        # sync with.
+
+        def _sync_trait_listener_deleted (ref, info):
+            for key, dic in info.items():
+                if key != '':
+                    for name, value in dic.items():
+                        if ref is value[0]:
+                            del dic[ name ]
+                            if len( dic ) == 0:
```
Comment (mdickinson) [inline traits/has_traits.py]: For simplicity and safety, can we unindent this a couple of levels?

Comment (mdickinson) [inline traits/has_traits.py]: Done in 094c55b31b4a690c01c4931827c9d80fe4fe81d1

File: traits/has_traits.py (hunk 1 above)
Comment (jvkersch) [inline traits/has_traits.py]: I'm not sure anymore this is so much of an issue. The traits change handler `_sync_trait_modified` will keep being triggered even after `self` doesn't sync with anything else anymore, but in that case it just iterates over an empty dictionary. I think I'll just delete my earlier comment.

**Comment by jvkersch** [comment]
@mdickinson Sorry for ignoring this PR for so long, and thanks for the refactoring! I removed what I thought is an unwarranted comment; let me know if you want to see further refactoring or other changes.

**Comment by mdickinson** [comment]
LGTM.  Waiting for Appveyor, then merging.

**Comment by jvkersch** [comment]
Everything is green, merging.

**Pull request merged**